src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,425 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
     }
 }
 
-jrawMonitorID
-debugMonitorCreate(char *name)
+/*
+ * A DebugRawMonitor is an encapsulation of a jrawMonitorID. It is used to
+ * store other information about the jrawMonitorID that is needed for ranked
+ * monitors support, such as the thread that currently owns the monitor.
+ */
+struct DebugRawMonitor {
+    jrawMonitorID monitor;
+    const char* name;
+    DebugRawMonitorRank rank;
+    volatile jthread ownerThread; // Access protected by dbgRawMonitor
+    int entryCount; // Only ownerThread should ever access this field
+};
+
+static DebugRawMonitor dbg_monitors[NUM_DEBUG_RAW_MONITORS];
+
+
+/*
+ * The dbgRawMonitor controls access to the DebugRawMonitor->ownerThread field.
+ * It must be held whenever modifying ownerThread and also must be held whenever
+ * accessing ownerThread unless we know the current thread already owns the monitor.
+ * Only the owner of the DebugRawMonitor will ever modify ownerThread. The main thing
+ * we are trying to protect from is a thread checking to see if it is the owner of
+ * the DebugRawMonitor, and having the ownerThread field change while doing this.
+ * The check is made using:
+ *    isSameObject(env, dbg_monitor->ownerThread, current_thread)
+ * We don't want ownerThread changing in the middle of this check since it can lead
+ * to a crash if the JNI handle gets freed. So this check is always protected by
+ * dbgRawMonitor as are modifications to ownerThread.
+ *
+ * There is also one other minor role that dbgRawMonitor plays. In verifyMonitorRank()
+ * we iterate over all monitors, and need to make sure the monitor we are checking has
+ * been initialized. If it hasn't been, then it is skipped. There can be a race and
+ * we catch it in the middle of intialization. Since dbgRawMonitor is held during
+ * verifyMonitorRank(), we should also hold it during monitor initialization.
+ */
+
+static jrawMonitorID dbgRawMonitor;
```

Comment:
As the monitor is used to synchronize access to dbg_monitors array, maybe rename it to something like dbg_monitors_lock?

Reply:
ok

Reply:
@alexmenkov Actually I don't think dbg_monitors_lock is a good name. It doesn't synchronize access to the dbg_monitors array, but rather to the ownerThread field of any given entry in the array. Maybe something like ownerThread_field_access_lock would be better, although that's getting a bit wordy, and maybe more specific than it needs to be (it would prevent other uses of the lock).